### ISSUE TITLE
TypeScript: Improve Definition

### ISSUE THREAD

**Issue by sgtoj** [testdouble/testdouble.js#238]
List of improvements I believe is needed for the TypeScript definition. I have already started working on this list. Love to hear feedback if you have any.

- [ ] Organize / Restructure  
- [ ] Add JSDoc for all functions
- [ ] Define `td.func()`
- [ ] Add generic types to `td.function()` and `td.func()`
- [ ] Define `td.constructor()`
- [ ] Add to `./test/src/typescript/test.ts` to reflect any additions.

**Comment by searls** [comment]
Any improvements are appreciated. Hey @duluca, you'd suggested helping out with our typescript story, want to look at this?

**Comment by sgtoj** [comment]
Updated OP with the link to the fork and branch.

**Comment by sgtoj** [comment]
PR #239

**Comment by duluca** [comment]
Will review. @sgtoj please lmk if there's anything specific I can help with. I'll leave any comments re: code on the pr

**Comment by sgtoj** [comment]
Disclaimer: I am not the best when it comes to communication so there may be the grammar or typos. I did a quick proof reading before committing. 

The jsdoc comments probably could use some addition or improvement from someone who has more experience with the framework and/or understanding behind the abstractions. Otherwise, it should be ready.

**Comment by duluca** [comment]
Sounds good. Will give it a shake tonight.

**Comment by albertogasparin** [comment]
Sorry to jump in, but with published typings I'm getting an error in VSCode (Typescript 2.3):
```js
const myMock = td.object(['run', 'search']);
td.when(myMock.run()).thenCallback(null);
// Error: Property 'run' does not exist on type 'DoubledObjectWithKey<"run" | "search">'.
```

Is it something this PR is going to solve? Or are you guys not getting this?

**Comment by searls** [comment]
@albertogasparin I suggest you attempt checking out the branch in #239 to find out for yourself

**Comment by albertogasparin** [comment]
Yep, was doing it. In fact, I'm still getting the same error even with the PR typings. Unfortunately I don't have much experience with Typescript, so I'm unable propose a fix

**Comment by sgtoj** [comment]
@albertogasparin I will look in to this issue. Hopefully, I will find a fix, add a test, and submit a new PR.

**Comment by sgtoj** [comment]
I fixed the issue and added a new TS test in #239. @albertogasparin @searls 

![Example](http://i.imgur.com/6MctLeg.png)

**Comment by albertogasparin** [comment]
@sgtoj Worked like a charm, thanks! 
Hope to see #239 merged soon 😉

**Comment by searls** [comment]
Good to hear. I'm chatting with @duluca tomorrow and want to have a chat with him about our TS approach more broadly before merging

**Comment by SimonSchick** [comment]
Any updates on this?

**Comment by sgtoj** [comment]
Let me know if there is anything I forgot to do or address.

**Comment by duluca** [comment]
@sgtoj as noted on PR #239, there are a couple of tasks due: 
TODO:
[ ] Update from master, resolve conflicts
[ ] Update PR and make sure CI checks pass

**Comment by sgtoj** [comment]
@duluca I believe we are ready.

**Comment by duluca** [comment]
Close, but no cigar. The test code must live in the regression/test.ts file.
